hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/HoodieHiveFileSliceReader.java

```diff
@@ -0,0 +1,190 @@
+      this.metaClient = HoodieTableMetaClient.builder().setConf(jobConf).setBasePath(basePath).build();
+      this.payloadProps.setProperty(HoodiePayloadProps.PAYLOAD_ORDERING_FIELD_PROP_KEY, metaClient.getTableConfig().getPreCombineField());
+      this.schema = new TableSchemaResolver(metaClient).getTableAvroSchema();
+    } catch (IOException e) {
+      throw new HoodieIOException("Failed to initialize metaclient for split: " + this.split);
+    } catch (Exception e) {
+      throw new HoodieException("Failed to get table avro schema.");
+    }
+  }
+
+  @Override
+  public FileSliceReader project(InternalSchema schema) {
+    // should we convert to avro Schema?
+    // TODO: implement me
+    return this;
+  }
+
+  @Override
+  public FileSliceReader project(Schema requiredSchema) {
+    String partitionFields = jobConf.get(hive_metastoreConstants.META_TABLE_PARTITION_COLUMNS, "");
+    List<String> partitioningFields =
+        partitionFields.length() > 0 ? Arrays.stream(partitionFields.split("/")).collect(Collectors.toList())
+            : new ArrayList<>();
+    this.writerSchema = HoodieRealtimeRecordReaderUtils.addPartitionFields(schema, partitioningFields);
+
+    List<String> projectionFields = HoodieRealtimeRecordReaderUtils.orderFields(jobConf.get(ColumnProjectionUtils.READ_COLUMN_NAMES_CONF_STR),
+        jobConf.get(ColumnProjectionUtils.READ_COLUMN_IDS_CONF_STR), partitioningFields);
+    Map<String, Schema.Field> schemaFieldsMap = HoodieRealtimeRecordReaderUtils.getNameToFieldMap(writerSchema);
+    this.readerSchema = HoodieRealtimeRecordReaderUtils.generateProjectionSchema(writerSchema, schemaFieldsMap, projectionFields);
+
+    return this;
+  }
+
+  @Override
+  public FileSliceReader pushDownFilters(Set<String> filters) {
+    String tableName = metaClient.getTableConfig().getTableName();
+
+    if (HoodieTableQueryType.INCREMENTAL.name().equals(readMode)) {
+      // TODO: construct predicate
+      // FilterPredicate predicate = constructHoodiePredicate(jobConf, tableName, split);
+      // LOG.info("Setting parquet predicate push down as " + predicate);
+      // ParquetInputFormat.setFilterPredicate(jobConf, predicate);
+    }
+
+    return this;
+  }
+
+  @Override
+  public FileSliceReader readingMode(HoodieTableQueryType queryType) {
+    this.readMode = queryType.name();
+    return this;
+  }
+
+  @Override
+  public Iterator<HoodieRecord> open(FileSlice fileSlice) {
+    try (HoodieFileReader baseFileReader = getReaderFactory(HoodieRecord.HoodieRecordType.HIVE).getFileReader(jobConf, split.getPath())) {
+      Iterator<HoodieRecord> baseRecordIterator = baseFileReader.getRecordIterator(readerSchema);
+      if (split instanceof RealtimeSplit) {
+        HoodieMergedLogRecordScanner logRecordScanner = getMergedLogRecordScanner((RealtimeSplit) split, jobConf, readerSchema);
+        while (baseRecordIterator.hasNext()) {
+          logRecordScanner.processNextRecord(baseRecordIterator.next());
```

Comment:
I don't think we'd be able to load whole file in memory. Instead, we should create an "merging iterator" going over the base-file and applying updates on the fly (similar to how Spark's one impl'd)